pkg/batcher/transport_message_batcher.go

```diff
@@ -0,0 +1,362 @@
+)
+
+type MessageArguments struct {
+	Metadata  map[string]string
+	Alias     string
+	Direction Direction
+	Protocol  protocol
+}
+
+type messageBatcher struct {
+	buffer      []byte
+	bufferIndex int
+
+	batchLenIndex  int
+	groupsLenIndex int
+	groupIndex     int
+
+	pipe        chan container
+	seqProvider sequenceProvider
+
+	batchSizeBytes uint
+	flushTimeout   time.Duration
+
+	router   message.Router
+	book     string
+	group    string
+	protocol protocol
+
+	done chan bool
+}
+
+func NewMessageBatcher(router message.Router, cfg MqMessageBatcherConfig) (MqBatcher[MessageArguments], error) {
+	flushTimeout := cfg.FlushMillis
+	if flushTimeout <= 0 {
+		flushTimeout = DEFAULT_FLUSH_TIME
+	}
+
+	maxBatchSizeBytes := cfg.BatchSizeBytes
+	if maxBatchSizeBytes <= 0 {
+		maxBatchSizeBytes = DEFAULT_BATCH_SIZE
+	}
+
+	flushDuration := time.Duration(flushTimeout) * time.Millisecond
+
+	batcher := messageBatcher{
+		buffer:         make([]byte, maxBatchSizeBytes),
+		pipe:           make(chan container, 1),
+		done:           make(chan bool, 1),
+		seqProvider:    sequenceProvider{},
+		batchSizeBytes: maxBatchSizeBytes,
+		flushTimeout:   flushDuration,
+		router:         router,
+		protocol:       cfg.Protocol,
+		group:          cfg.Group,
+		book:           cfg.Book,
+	}
+	batcher.reset() // reset flushing time for first message
+	go batcher.flushingRoutine()
+	return &batcher, nil
+}
+
+func (b *messageBatcher) Send(data []byte, args MessageArguments) (int, error) {
+	dataLength := len(data)
+	if dataLength <= 0 {
+		return 0, nil
+	}
+
+	overhead := b.computeOverhead(data, args)
+	if overhead > int(b.batchSizeBytes) {
+		return overhead, fmt.Errorf("too large message of data %d, serialized size %d, max %d", dataLength, overhead, b.batchSizeBytes)
+
+	}
+
+	b.pipe <- container{data, args, overhead}
+
+	return overhead, nil
+}
+
+func (b *messageBatcher) Close() error {
+	log.Info().Msg("Closing message batcher")
+	close(b.pipe)
+	<-b.done
+	return nil
+}
+
+func (b *messageBatcher) flushingRoutine() {
+	log.Info().Msg("Flushing routine started")
+	defer func() {
+		b.done <- true
+	}()
+	for {
+		timer := time.After(b.flushTimeout)
+
+		select {
+		case item, ok := <-b.pipe:
+			if !ok {
+				log.Debug().Msg("Flushing messages by pipe complete")
+				b.flush()
+
+				log.Info().Msg("Flushing routine stopped")
+				return
+			}
+			if b.bufferIndex+item.overhead > cap(b.buffer) {
+				log.Debug().Msg("Flushing messages by buffer size")
+				b.flush()
+			}
+
+			b.write(item.data, item.args.Metadata, item.args.Alias, item.args.Direction, item.args.Protocol)
+			if b.bufferIndex == cap(b.buffer) {
+				log.Debug().Msg("Flushing messages by buffer is full")
+				b.flush()
+			}
+		case <-timer:
+			log.Debug().Msg("Flushing messages by timer is over")
+			b.flush()
+		}
+	}
+
+}
+
+func (b *messageBatcher) write(data []byte, metadata map[string]string, alias string, direction Direction, protocol string) {
+	dst := b.buffer[b.bufferIndex:]
```

Comment:
note: You can use https://github.com/th2-net/transport-go instead. It should simplify this part and should not affect the performance much. In this case, we probably should modify the `Encoder` in that module to allow checking the available space for a new message

Reply:
I think about using `Encoder` instead of our buffer. We can add the methods to checking space of encoding 
* SizeAfterEncodeRaw
* SizeAfterEncodeParsed

Reply:
Okay, let's see what will it look like